### PR TITLE
Issue #440

### LINKED ISSUE #440: Update Service Legacy Code
**Is your feature request related to a problem? Please describe.**
Currently the service legacy code is not up to the current spec.

**Describe the solution you'd like**
Update the code to the current spec.

**(Optional) Include a BDD test that would satisfy the expected behaviour of the new feature**
All BDD need to test the same functionality as the legacy and pass

### PR DESCRIPTION
**Pull request by andrewnguyen22**
Closes #440

### CODE REVIEW

File: tests/bdd/servicing/service_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+					Expect(relayMissingBlockchain.IsValid(hostedBlockchains)).To(Equal(service.EmptyBlockchainError))
+				})
+			})
+			Context("Unsupported blockchain", func() {
+				It("should return an unsupported blockchain error", func() {
+					Expect(relayUnsupportedBlockchain.IsValid(hostedBlockchains)).To(Equal(service.UnsupportedBlockchainError))
+				})
+			})
+		})
+		Context("Invalid payload", func() {
```
Comment (luyzdeleon) [inline tests/bdd/servicing/service_test.go:24]: I believe `payload` should be optional. Or maybe we should restructure the Relay data structure.

File: tests/bdd/servicing/service_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			})
+			Context("Invalid client increment counter signature", func() {
+				It("should return an invalid incremnt counter signature error", func() {
+					Expect(relayInvalidICSignature.IsValid(hostedBlockchains).Error()).To(ContainSubstring(service.InvalidICSignatureError.Error()))
+				})
+			})
+		})
+		Context("Valid relay data", func() {
+			It("should return no error", func() {
+				Expect(validRelay.IsValid(hostedBlockchains)).To(BeNil())
```
Comment (luyzdeleon) [inline tests/bdd/servicing/service_test.go]: `IsValid` should return a bool and an error. Or maybe be something like: `validRelay.validate(hostedBlockchains)` and that way returning just an `error` makes more sense.

File: tests/bdd/servicing/service_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			})
+		})
+	})
+	PDescribe("Relay execution", func() {})
```
Comment (luyzdeleon) [inline tests/bdd/servicing/service_test.go:82]: Should we add a `PDescribe("Relay caching", func() {})` to include caching the work done?

File: x/service/incrementCounter.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	"errors"
+	"github.com/pokt-network/pocket-core/crypto"
+)
+
+type IncrementCounter struct {
+	Counter   int              `json:"counter"`
+	Signature crypto.Signature `json:"signature"`
+}
+
+func (ic IncrementCounter) IsValid(clientPubKey string, messageHash []byte) error {
```
Comment (luyzdeleon) [inline x/service/incrementCounter.go]: Rename this to `Validate` so it makes more syntactical sense, or return both a bool and an error.

File: x/service/relay.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+type Relay struct {
+	Blockchain       ServiceBlockchain `json:"blockchain"`
+	Payload          ServicePayload    `json:"payload"`
+	ServiceToken     ServiceToken      `json:"token"`
+	IncrementCounter IncrementCounter  `json:"incrementCounter"`
+}
+
+func (r Relay) IsValid(hostedBlockchains ServiceBlockchains) error {
```
Comment (luyzdeleon) [inline x/service/relay.go]: Rename this to `Validate` so it makes more syntactical sense, or return both a bool and an error.

File: x/service/relay.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+func (r Relay) Execute(hostedBlockchains ServiceBlockchains) (string, error) {
+	if err := r.IsValid(hostedBlockchains); err != nil {
+		return "", err
+	}
+	chainURL, err := hostedBlockchains.GetChainURL(r.Blockchain.String())
+	if err != nil {
+		return "", err
+	}
+	result, err := executeHTTPRequest(r.Payload.Data.Bytes(), chainURL, r.Payload.Method)
```
Comment (luyzdeleon) [inline x/service/relay.go:50]: I think this should be done according to a simplified interface that depending on what's contained on the `Relay` determines the network protocol to route this to.

File: x/service/serviceToken.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"encoding/hex"
+	"github.com/pokt-network/pocket-core/crypto"
+	"github.com/pokt-network/pocket-core/types"
+)
+
+type ServiceToken types.AAT
+
+func (st ServiceToken) IsValid() error {
```
Comment (luyzdeleon) [inline x/service/serviceToken.go]: Same as above.

File: x/service/serviceToken.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	// }
+	return nil
+}
+
+func (st ServiceToken) Hash() []byte {
+	// TODO return hash of the amino encoding of service token
+	return crypto.SHA3FromString(st.AATMessage.ApplicationPublicKey + st.AATMessage.ClientPublicKey) // temporary
+}
+
+func (st ServiceToken) VersionIsValid() error {
```
Comment (luyzdeleon) [inline x/service/serviceToken.go]: Same as above.

File: x/service/serviceToken.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	if !st.Version.IsIncluded() {
+		return MissingTokenVersionError
+	}
+	if !st.Version.IsSupported() {
+		return UnsupportedTokenVersionError
+	}
+	return nil
+}
+
+func (st ServiceToken) MessageIsValid() error {
```
Comment (luyzdeleon) [inline x/service/serviceToken.go]: Same as above.

File: x/service/serviceToken.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		return MissingApplicationPublicKeyError
+	}
+	// todo pub key format verification
+	if len(st.AATMessage.ClientPublicKey) == 0 {
+		return MissingClientPublicKeyError
+	}
+	return nil
+}
+
+func (st ServiceToken) SignatureIsValid() error {
```
Comment (luyzdeleon) [inline x/service/serviceToken.go]: Same as above.

**Review by luyzdeleon — CHANGES_REQUESTED** [review]

File: tests/bdd/servicing/service_test.go (hunk 1 above)
Comment (andrewnguyen22) [reply tests/bdd/servicing/service_test.go:24]: I don't think there is a situation where the payload would be optional

File: tests/bdd/servicing/service_test.go (hunk 2 above)
Comment (andrewnguyen22) [reply tests/bdd/servicing/service_test.go]: Okay, that works with me!

File: tests/bdd/servicing/service_test.go (hunk 3 above)
Comment (andrewnguyen22) [reply tests/bdd/servicing/service_test.go:82]: I thought we were pushing `caching` to the infra level? If you are referring to relay batching then see PDescribe("Relay Batching")

File: x/service/incrementCounter.go (hunk 4 above)
Comment (andrewnguyen22) [reply x/service/incrementCounter.go]: Cool

File: x/service/relay.go (hunk 5 above)
Comment (andrewnguyen22) [reply x/service/relay.go]: Cool

File: x/service/relay.go (hunk 6 above)
Comment (andrewnguyen22) [reply x/service/relay.go:50]: Yes agreed, this is a placeholder

File: x/service/serviceToken.go (hunk 7 above)
Comment (andrewnguyen22) [reply x/service/serviceToken.go]: Cool

File: x/service/serviceToken.go (hunk 8 above)
Comment (andrewnguyen22) [reply x/service/serviceToken.go]: Cool

File: x/service/serviceToken.go (hunk 9 above)
Comment (andrewnguyen22) [reply x/service/serviceToken.go]: Cool

File: x/service/serviceToken.go (hunk 10 above)
Comment (andrewnguyen22) [reply x/service/serviceToken.go]: Cool

**Review by luyzdeleon — DISMISSED** [review]

**Review by luyzdeleon — APPROVED** [review]

**Pull request merged**